### DISCUSSION TITLE
Array of objects is converted to array of arrays on next request

### DISCUSSION THREAD

**Discussion by hulkur** [livewire/livewire#3041] [7. Bugs (archived)]
### Description

I initialize a variable from DB as array of objects (QueryBuilder, get()->toArray()).
On next request the variable comes up as array of arrays.

### Stripped-down, copy-pastable code snippets

in component: called once on user input in updated method (`updatedFoo()`) 

```
$orgs = Organization::select(['id', 'type', 'name'])
            ->toBase()
            ->get()
            ->toArray();
```

in blade:

```
@foreach($orgs as $o)
     <option value="{{ $o->id }}">{{ $o->name }}</option>
@endforeach
```

First request it works, next request gives error "Attempt to read property "id" on array"

**Context**
- Livewire version: 2.4.2
- Laravel version: 8.37.0
- Alpine version: 2.8.2

**Comment by calebporzio** [comment]
What happens if you get rid of the "toArray' at the end?

**Comment by hulkur** [comment]
first request: collection of objects
next request: collection of arrays

**Comment by hulkur** [comment]
minimal test: 

component:

```
public $test = null;

public function mount()
{
    if (is_null($this->test)) {
        $a = new \stdClass(); $a->a = 1;
        $b = new \stdClass(); $b->a = 2;
        $this->test = collect([$a, $b]);
    }
}
```

in blade add: `@php dump($test) @endphp`

Variable $test is not touched anywhere else. 
On first request it is dumped as 

```
Illuminate\Support\Collection {#1607 ▼
  #items: array:2 [▶
    0 => {#1613 ▶
      +"a": 1
    }
    1 => {#1608 ▶
      +"a": 2
    }
  ]
}
```

On next request after some update:

```
Illuminate\Support\Collection {#632 ▼
  #items: array:2 [▶
    0 => array:1 [▶
      "a" => 1
    ]
    1 => array:1 [▶
      "a" => 2
    ]
  ]
}
```

**Comment by joshhanley** [comment]
@hulkur I'd expect that behaviour to happen as Livewire can't rehydrate stdObjects, as such they just end up as arrays. The reason for that is Laravel request object, when it runs json_decode, forces json objects to become arrays. See this issue laravel/framework#443

But it can do Eloquent collections without any issues. 

So my question is, why are you using `->toBase()` and `->toArray()`? Why don't you just use the following?

`$orgs = Organization::select(['id', 'type', 'name'])->get();`

This is one issue I found this week when helping someone else, was when using `->toBase()` and `->toArray()` together, is the base query will construct a collection of stdObjects, but when you call to array, it only converts the collection to an array, but won't convert the stdObjects to arrays (which is unexpected in my opinion). Hence why you are getting a mismatch.

If you do need it for whatever reason, there are two ways around it currently. Either convert all stdObjects to arrays as part of the query, so before and after the request you are only working with arrays; or you can use a hydrate hook to manually convert the rehydrated arrays back to stdObjects.

This is the first option
```php
$orgs = Organization::select(['id', 'type', 'name'])
            ->toBase()
            ->get()
            ->map(function($org) {
                return (array) $org;
            })
            ->toArray();
```

@calebporzio I've seen quite a few people run into issues with stdObjects. I wonder whether that is something we should support serialising and deserialising? Especially in a scenario like above with a custom select?

Would involve inspecting a collection to see if it contains stdObjects and keeping track of that as another meta type. Converting from array to stdObject is as simple as putting `(object)` cast in front of the array.

Hope this helps!

**Comment by hulkur** [comment]
this is exactly what I ended up doing.

```
            ->map(function($org) {
                return (array) $org;
            })
```

I use `toBase` when I don't need the full object but only few fields from it. Usually it's id and name for select.

**Comment by calebporzio** [comment]
Yeah, interesting. Supporting more non js data types of a bit of a never ending persuit.

I'd be interested in coming up with a better, recursive system that allows plugging in extra data type support.

This would be something for v3 I think.

I'm hesitant to keep tacking on to the current system.

Thoughts?

**Comment by nuernbergerA** [comment]
@calebporzio 

We could do some heavy lifting there and evaluate each variable on init and categorize them.

For arrays that contain non-js primitives, we can create a blueprint of the structure.

Maybe we can open a new feature request issue with a list of what is supported and what is missing?

**Comment by torian257x** [comment]
Maybe laravel has something to properly serialize models etc?

I saw Laravel has a trait
`trait SerializesModels` 

it gets used for jobs in queues... 

What are the caveats? I don't know honestly. There must be some.

This should be explained in the docs, but a search for serialize returns nothing. It's not logical that an array of models becomes an array of arrays - what feels like randomly. 

I lost quite some time to understand waddup, which I regret losing.

**Comment by torian257x** [comment]
@calebporzio  you have the same problem by just keeping state client side. That is very dangerous actually...
first is encryption yes, but before that, you should be able to keep the state server side
when you do that, you should implement as well a way to hydrate models etc

https://stackoverflow.com/a/32933875/533426
to have consistent state

**Comment by redpawn96** [comment]
Hi guys, what happened to this issue? Is there any workaround? 

My temporary solution is to create a public variable that holds the Eloquent collection  every render.. In this way the model will not be turned into array of arrays in the next request..

**Comment by jazz-max** [comment]
Hi. I solved this issue by write custom wireable collection class for store simple collection of objects.


```
class WireableCollection extends Collection implements Wireable
{

    public function __construct ($items)
    {
        return parent::__construct ( $items );
    }

    public static function fromLivewire ($value): static
    {
        $value = json_decode ( json_encode ( $value ) );

        return new static( $value );
    }

    public function toLivewire (): array
    {
        return $this->items;
    }
}
```

**Comment by Dimimo** [comment]
My solution is to keep an array called $ids; as a temporarily substitute for the $tags collection (as a tag selector for articles). When a choice is made in the search dropdown, I trigger this event:

```PHP
   public function selectChoice($id)
    {
        //workaround to avoid previous collections convert to arrays
        if ($tag = Tag::find($id)) {
            $this->ids[] = $tag->id;
            $this->ids = array_unique($this->ids);
        }
        $this->tags = Tag::whereIn('id', $this->ids)->orderBy('name')->get();
   }
```
At re-render $tags is now a full eloquent collection.

**Comment by JavierMarcosAle** [comment]
I was having this problem when consuming an API that returns me an array of objects.

One solution I got on another forum was to put the method consuming the API in **render()** instead of **mount()**, and it did work. The problem with this solution is that **render** causes the API to be called again, which is expected behavior, but a bad user experience.

I have solved it by doing what the previous comments indicate. Convert the object to an array. Now I can make the call on **mount()** with no problem

```
$array = [];
foreach ($data as $objeto) {
    $array[] = get_object_vars($objeto);
}

return $array;
```

**Comment by SpyridonLaz** [comment]
I am working on a  huge project where we execute SQL Stored Procedures through DB facade instead of Eloquent and this has become a major drawback for us since re-executing stored procedures during hydration kinda defeats the purpose of having them in the first place( for faster db calls).

For example :

```php
    public function mount()
          {
              $this->data = collect(DB::select('exec FN_Get_Company_Vehicles'));
          }
```
then after update:

```php
    public function hydrate(){
        $this->data = collect(DB::select('exec FN_Get_Company_Vehicles'));
     
    }

```

Now imagine having dozens of DB::select at the same time. Obviously this is not a practical solution.

Now, in blade file:

```blade
      <td class="px-3 py-1 text-sm text-right"> {{ $item-> PersonalNr ??$item['PersonalNr'] }} </td>
```
and that is to catch both the first render and all the consecutive ones.

@calebporzio  My thoughts are that Livewire could somehow serialize/deserialize DB queries as well as Models which is  intergrated in laravel and used in many projects where transaction speed is important. It would be so , so pain relieving.

**Reply by jamessoncardozo** [reply]
I'm facing the same problem! The first item off collection is rendered but, when I add another item, somthing change my first item in a array! >>> More details >>> https://stackoverflow.com/q/77215874/20947756?stw=2
